src/main/java/com/amazon/opendistroforelasticsearch/ad/rest/handler/AbstractAnomalyDetectorActionHandler.java

```diff
@@ -0,0 +1,670 @@
+                if (errorMsg != null) {
+                    listener.onFailure(new OpenSearchStatusException(errorMsg, indexResponse.status()));
+                    return;
+                }
+                listener
+                    .onResponse(
+                        new IndexAnomalyDetectorResponse(
+                            indexResponse.getId(),
+                            indexResponse.getVersion(),
+                            indexResponse.getSeqNo(),
+                            indexResponse.getPrimaryTerm(),
+                            detector,
+                            RestStatus.CREATED
+                        )
+                    );
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                logger.warn("Failed to update detector", e);
+                if (e.getMessage() != null && e.getMessage().contains("version conflict")) {
+                    listener
+                        .onFailure(
+                            new IllegalArgumentException("There was a problem updating the historical detector:[" + detectorId + "]")
+                        );
+                } else {
+                    listener.onFailure(e);
+                }
+            }
+        });
+    }
+
+    protected void onCreateMappingsResponse(CreateIndexResponse response, boolean indexingDryrun) throws IOException {
+        if (response.isAcknowledged()) {
+            logger.info("Created {} with mappings.", ANOMALY_DETECTORS_INDEX);
+            prepareAnomalyDetectorIndexing(indexingDryrun);
+        } else {
+            logger.warn("Created {} with mappings call not acknowledged.", ANOMALY_DETECTORS_INDEX);
+            listener
+                .onFailure(
+                    new OpenSearchStatusException(
+                        "Created " + ANOMALY_DETECTORS_INDEX + "with mappings call not acknowledged.",
+                        RestStatus.INTERNAL_SERVER_ERROR
+                    )
+                );
+        }
+    }
+
+    protected String checkShardsFailure(IndexResponse response) {
+        StringBuilder failureReasons = new StringBuilder();
+        if (response.getShardInfo().getFailed() > 0) {
+            for (ReplicationResponse.ShardInfo.Failure failure : response.getShardInfo().getFailures()) {
+                failureReasons.append(failure);
+            }
+            return failureReasons.toString();
+        }
+        return null;
+    }
+
+    protected void validateAnomalyDetectorFeatures(String detectorId, boolean indexingDryrun) throws IOException {
+        if (anomalyDetector != null
+            && (anomalyDetector.getFeatureAttributes() == null || anomalyDetector.getFeatureAttributes().isEmpty())) {
+            checkADNameExists(detectorId, indexingDryrun);
+            return;
+        }
+        String error = RestHandlerUtils.validateAnomalyDetectorFeaturesConfig(anomalyDetector, maxAnomalyFeatures);
```

Comment:
For now I will consider changing to better names instead of combining them together. Such validation against features includes 2 parts: 1) configuration error, or something like "syntax" error, such as over 5 features, duplicate feature name; and 2) runtime error, such "sum" aggregation on keyword field. 

`validateAnomalyDetectorFeaturesConfig` is for "syntax" error checking, and the rest part of `validateAnomalyDetectorFeatures` is for "runtime" error.

The reason of changing to better name is because `validateAnomalyDetectorFeaturesConfig` is also used by Preview API now. Changing Preview API to use `validateAnomalyDetectorFeatures` to cover both type of errors may return errors in case of "runtime" error found. In error case, current Preview API returns empty result, and it can be handled by front-end; if we change to use `validateAnomalyDetectorFeatures`, front-end won't be able to handle the result> Thus, it requires more effort on both front-end/backend to do the refactoring if we combine them into 1 method. I think this can be an improvement area for us in the future. 

Regarding the name changing, what about changing `validateAnomalyDetectorFeaturesConfig` to `checkAnomalyDetectorFeaturesSyntax`, and keep `validateAnomalyDetectorFeatures` ?

Reply:
Sounds good.  Please also add your explanation to comments.